### DISCUSSION TITLE
Modbus Bridge RTU -> TCP

### DISCUSSION THREAD

**Discussion by makru83** [eModbus/eModbus#329] [Q&A] [answered]
I have 3 "Pego 200" Fridge Controllers, which I would like to integrate into our Home automation System, which has a Modbus TCP/IP functionality build in.

I used the example code "Wifi-RTU_Bridge", changed Wifi Settings, and Baud Rate of the RTU Device, and sent it to a esp32.
I am using a XY-017 Adaptor connected to Gpio16 & 17.
For testing I am using a Windows Computer with a RS485 Usb Stick with trial Modbus Slave Tool.
On a second Computer I use Modbus Poll to connect to the exp32.

The Windows Slave is ID:1, and on the Modbus Poll I try to connect to ID:4

I tried a lot of different Variations, but I am unable to get any Data with the Modbus Poll Software.

Is the Wifi-RTU_Bridge Example actually for this Kind of scenario?

**Comment by Miq1** [comment]
Watch out: GPIOs 16 and 17 are switched in the example code - RX and TX are reversed to the usual setup. This is due some odd wiring I had initially... 🙄 

Otherwise the example is what you want to have. You need to use ``attachServer`` to bind each server on the RTU bus to a dedicated server ID the bridge will listen to. The bridge server ID can be the same the RTU server is using if you like.

An external WiFi client then will have to request the bridge for that bridge server ID, the bridge will forward the request to the attached server and bring back any response it receives to the WiFi client.

**Comment by makru83** [comment]
Thanks, unfortunatly I don't get it to work. I already exchanged the rs485 Adapter but no luck.
The Txd light on the Adapter lights up about every second.

This is the code I used:

```
// =================================================================================================
// eModbus: Copyright 2020 by Michael Harwerth, Bert Melis and the contributors to ModbusClient
//               MIT license - see license.md for details
// =================================================================================================
// Includes: <Arduino.h> for Serial etc., WiFi.h for WiFi support
#include <Arduino.h>
#include "HardwareSerial.h"
#include <WiFi.h>

// Modbus bridge include
#include "ModbusBridgeWiFi.h"
// Modbus RTU client include
#include "ModbusClientRTU.h"

#ifndef MY_SSID
#define MY_SSID "my-wifi"
#endif
#ifndef MY_PASS
#define MY_PASS "my-password"
#endif

char ssid[] = MY_SSID;                     // SSID and ...
char pass[] = MY_PASS;                     // password for the WiFi network used
uint16_t port = 502;                       // port of modbus server

// Create a ModbusRTU client instance
ModbusClientRTU MB;

// Create bridge
ModbusBridgeWiFi MBbridge;

// Setup() - initialization happens here
void setup() {
// Init Serial monitor
  Serial.begin(115200);
  while (!Serial) {}
  Serial.println("__ OK __");

// Init Serial2 conneted to the RTU Modbus
// (Fill in your data here!)
  RTUutils::prepareHardwareSerial(Serial2);
  Serial2.begin(9600, SERIAL_8N1, GPIO_NUM_16, GPIO_NUM_17);

// Connect to WiFi
  WiFi.begin(ssid, pass);
  delay(200);
  while (WiFi.status() != WL_CONNECTED) {
    Serial.print('.');
    delay(1000);
  }
  IPAddress wIP = WiFi.localIP();
  Serial.printf("IP address: %u.%u.%u.%u\n", wIP[0], wIP[1], wIP[2], wIP[3]);

// Set RTU Modbus message timeout to 2000ms
  MB.setTimeout(2000);
// Start ModbusRTU background task on core 1
  MB.begin(Serial2, 1);

// Define and start WiFi bridge
// ServerID 4: Server with remote serverID 1, accessed through RTU client MB
//             All FCs accepted, with the exception of FC 06
  MBbridge.attachServer(1, 1, ANY_FUNCTION_CODE, &MB);
  
// ServerID 5: Server with remote serverID 4, accessed through RTU client MB
//             Only FCs 03 and 04 accepted
  MBbridge.attachServer(2, 2, READ_HOLD_REGISTER, &MB);
  MBbridge.addFunctionCode(2, READ_INPUT_REGISTER);

// Check: print out all combinations served to Serial
  MBbridge.listServer();

// Start the bridge. Port 502, 4 simultaneous clients allowed, 600ms inactivity to disconnect client
  MBbridge.start(port, 4, 600);

  Serial.printf("Use the shown IP and port %d to send requests!\n", port);

// Your output on the Serial monitor should start with:
//      __ OK __
//      .IP address: 192.168.178.74
//      [N] 1324| ModbusServer.cpp     [ 127] listServer: Server   4:  00 06
//      [N] 1324| ModbusServer.cpp     [ 127] listServer: Server   5:  03 04
//      Use the shown IP and port 502 to send requests!
}

// loop() - nothing done here today!
void loop() {
  delay(10000);
}
```


Can I add some sort of debugging feature, so I could check in the serial Monitor?
![modbus](https://github.com/eModbus/eModbus/assets/2911058/737e012a-c7bc-450d-a9c8-e7ed08d67885)
![20240104_210829](https://github.com/eModbus/eModbus/assets/2911058/dc8de21e-16ff-43cc-985c-7b5487a11cc6)

**Reply by Miq1** [reply]
Side note: you are missing the biasing resistors on the RS485 side.

You can edit ``Logging.h`` to set the default ``LOG_LEVEL`` to ``LOG_LEVEL_VERBOSE``. Then you will get lots of debugging output on the Serial Monitor you can post here for me to analyze.

**Reply by makru83** [reply]
Thank you. I got it working now.
After changing the Timeout, I did get Messages in the Log, and I have seen that there are results coming.
So I changed the RTU Slave Parity to none, now it's working. I still get quite some Errors in the Modbus Poll Software, but this might be because of the missing resistor.

I checked the Documentation, but I cannot find where can I set the RTU Details of Data length, Stop Bits and Parity.
In the Example I can see "Serial_8N1", but no explanation to it.

**Reply by Miq1** [reply, marked as answer]
This has nothing to do with RTU but is ESP32 UART basics. Have a look at the ``HardwareSerial`` documentation for the MCU.
